File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool result = true;
    int helper(TreeNode* root) {
        if (!root) return 0;
        
        int left = helper(root->left);
        int right = helper(root->right);

        if (abs(left - right) > 1) {
            result = false;
        }

        return 1 + max(left, right);
    }
    bool isBalanced(TreeNode* root) {
        helper(root);

        return result;
    }
};

File: Data Structures & Algorithms/sqrtx/submission-0.py
class Solution:
    def mySqrt(self, x: int) -> int:
        left, right = 0, x
        while left <= right:
            middle = (left + right) // 2
            if middle * middle  > x:
                right = middle - 1
            elif middle * middle < x:
                left = middle + 1
            else:
                return middle
        return left - 1
